libavcodec/qsvenc.c

```diff
@@ -825,6 +840,47 @@ static int init_video_param(AVCodecContext *avctx, QSVEncContext *q)
         q->extparam_internal[q->nb_extparam_internal++] = (mfxExtBuffer *)&q->extvsi;
     }
 
+    if (q->dual_gfx) {
+#if QSV_HAVE_HE
+        if(q->param.mfx.CodecId != MFX_CODEC_AVC && q->param.mfx.CodecId != MFX_CODEC_HEVC) {
+            av_log(avctx, AV_LOG_WARNING, "Not supported encoder for dual GFX mode. "
+                                          "Supported: h264_qsv and hevc_qsv \n");
+        }
+        if(q->param.mfx.RateControlMethod != MFX_RATECONTROL_VBR &&
+           q->param.mfx.RateControlMethod != MFX_RATECONTROL_CQP &&
+           q->param.mfx.RateControlMethod != MFX_RATECONTROL_ICQ) {
+            av_log(avctx, AV_LOG_WARNING, "Not supported BRC for dual GFX mode. "
+                                          "Supported: VBR, CQP and ICQ \n");
+        }
+        if(q->param.mfx.LowPower != MFX_CODINGOPTION_ON) {
+            av_log(avctx, AV_LOG_WARNING, "Dual GFX mode supports only low-power encoding mode \n");
+        }
+        if((q->param.mfx.CodecId == MFX_CODEC_AVC  && q->param.mfx.IdrInterval != 0) ||
+           (q->param.mfx.CodecId == MFX_CODEC_HEVC && q->param.mfx.IdrInterval != 1)) {
+            av_log(avctx, AV_LOG_WARNING, "Dual GFX mode requires closed GOP (IdrInterval = 0) \n");
+        }
+        if(q->param.mfx.GopPicSize > 30) {
+            av_log(avctx, AV_LOG_WARNING, "For better performance in dual GFX mode GopPicSize must be <= 30 \n");
+        }
+        if(q->param.AsyncDepth < 30) {
+            av_log(avctx, AV_LOG_WARNING, "For better performance in dual GFX mode AsyncDepth must be >= 30 \n");
+        }
+
+        mfxIMPL impl;
+        MFXQueryIMPL(q->session, &impl);
+        if (MFX_IMPL_VIA_MASK(impl) != MFX_IMPL_VIA_D3D11) {
+            av_log(avctx, AV_LOG_ERROR, "Dual GFX mode requires D3D11VA \n");
```

Comment:
I believe in the case of DX9 it will fail even for ADAPTIVE mode? Right?
Should we get into all of these conditions for VAAPI on Linux if not supported?
Maybe better to check as first condition just after #if QSV_HAVE_HE 844 line and exit if not supported?

Reply:
> I believe in the case of DX9 it will fail even for ADAPTIVE mode? Right?

Yes, rigth

> Should we get into all of these conditions for VAAPI on Linux if not supported?

I suppose that with VAAPI on Linux we won't get into all of these conditions, because:
```
if (q->dual_gfx) {
#if QSV_HAVE_HE
         ...
#else
        av_log(avctx, AV_LOG_WARNING, "Dual GFX mode is supported only on Windows and requires oneVPL \n");
}
```
and
```
#if defined(_WIN32) || defined(__CYGWIN__)
#define QSV_HAVE_HE     QSV_ONEVPL
#else
#define QSV_HAVE_HE     0
#endif
```

> Maybe better to check as first condition just after #if QSV_HAVE_HE 844 line and exit if not supported?

But even so it probably makes sense (I don't know if this is checked for VPL). I will fix. I think we can also return an error for:
```
av_log(avctx, AV_LOG_WARNING, "Dual GFX mode is supported only on Windows and requires oneVPL \n");
```

Reply:
Done